Accessibility and Usability Optimisations

telmiger opened the pull request:
As discussed with Josiah … 
https://github.com/tgrosinger/tw5-checklist/issues/29#issuecomment-304064258

The demo is updated here: https://tid.li/tw5/test/checklist.html 

A big step for this wonderful plugin :)

telmiger committed:
Configuration is available now

Misleading text deleted.

telmiger committed:
Several optimisations - not done yet

Inspired by the great article here: https://inclusive-components.design/a-todo-list/ I tried to make the list more accessible and more structured.
Still missing: Separation of list and input element (make a form!) as well as clear-all button (make a button!)

telmiger committed:
New type, new styles

Removed type text/css so we can use WikiText variables and filters … moved logic for strike-through font choice to here.
Added styling for label and for items with more than one line of text.

telmiger committed:
Accessibility improvements, moved styling to css

labels and buttons added, decoration is removed

telmiger committed:
Adaption for label, comments added

One function was commented out: the pencil is no longer replaced by a (non-functional) checkbox.

telmiger committed:
Reload > Rerender and style for this tiddler integrated

A special style for this tiddler only was moved from the stylesheet and integrated here. The style was renamed as the name was misleading: the list needs rerendering (not a page reload) to show configuration changes. The text was adapted accordingly.

telmiger committed:
Refactoring, restructuring, additions

Added styles for visually hidden content (for screenreaders), for placeholder, for new elements (e.g. buttons) - used color variables to keep the style in line with the user’s template.

tgrosinger commented on tiddlers/usage.tid:
```diff
@@ -39,7 +39,3 @@ delimited) checklist items will be displayed as a single checklist.
 List items support any inline formatting rules. This means you can create links,
 bold, or anything else found on the
 [[Formatting in WikiText|http://tiddlywiki.com/static/Formatting%2520in%2520WikiText.html]] guide.
-
```
This should be addressed as part of #30. I would like to not remove this right now because it is a good reminder that it needs to be replaced with a link to the actual documentation.

tgrosinger commented on js/checkbox-wikirule.js:
```diff
@@ -21,30 +21,23 @@ exports.init = function(parser) {
 exports.shouldShowClearAll = function() {
     var configWidgetTitle = "$:/plugins/tgrosinger/tw5-checklist/Configuration";
     var configWidgetFields = $tw.wiki.getTiddler(configWidgetTitle).fields;
-
```
minimum diff please

tgrosinger commented on js/checkbox-wikirule.js:
```diff
@@ -21,30 +21,23 @@ exports.init = function(parser) {
     return (showClearAll === "true");
 }
 
+
```
minimum diff please

tgrosinger commented on js/checkbox-wikirule.js:
```diff
@@ -21,30 +21,23 @@ exports.init = function(parser) {
-Retrieve the configuration state indicating if items should be striken
+Retrieve the configuration state of the clear all button
 */
-exports.shouldStrikeChecked = function() {
```
Why was this removed?

telmiger replied:
We can check this directly from the CSS now, see below. As stirkethrough is a styling issue, I thought this would be the right place.

tgrosinger commented on js/checkbox-wikirule.js:
```diff
@@ -79,40 +72,54 @@ exports.parse = function() {
         };
         if (match[1] === "x" || match[1] === "X") {
             checkbox.attributes.checked = {type: "boolean", value: true};
-            if (this.shouldStrikeChecked()) {
```
Why was this removed?

telmiger replied:
The stylesheet took over this check. See below.

tgrosinger commented on js/checkbox-wikirule.js:
```diff
@@ -79,40 +72,54 @@ exports.parse = function() {
         }
 
-        var removeicon = {
+        var itembody = {
```
Can you move this back up above checkbox? Minimum diff please

tgrosinger commented on js/checkbox-wikirule.js:
```diff
@@ -79,40 +72,54 @@ exports.parse = function() {
-            type: "element",
-            tag: "span",
-            attributes: {
-                class: {type: "string", value: "checklistitem-body"}
```
if checklistitem-body is no longer used it should be removed as part of this commit. Each commit should be a functional intermediate state.

telmiger replied:
This was still in use, I had replaced the element with a label instead of a span. But you are right, we can remove it and target the label element without a class.

tgrosinger commented on js/checkbox-wikirule.js:
```diff
@@ -79,40 +72,54 @@ exports.parse = function() {
         if (match[1] === "x" || match[1] === "X") {
             checkbox.attributes.checked = {type: "boolean", value: true};
-            if (this.shouldStrikeChecked()) {
-                itembody.attributes.class.value += ", checkedchecklistitem-body";
```
Same as previous comment about checklistitem-body, but for checkedchecklistitem-body.

tgrosinger commented on tiddlers/stylesheet.tid:
```diff
@@ -1,6 +1,5 @@
 title: $:/plugins/tgrosinger/tw5-checklist/stylesheet
 tags: $:/tags/Stylesheet
-type: text/css
```
Why was this removed?

telmiger replied:
This type does not allow wikitext interpretation, which is used for a) transclusion of color values and b) checking the strikethrough-setting. (No type ist the same as text/vnd.tiddlywiki I believe.)

tgrosinger commented on tiddlers/stylesheet.tid:
```diff
@@ -13,24 +12,33 @@ ul.checklist > li {
 
-span.checkedchecklistitem-body {
-    color: #ddd;
+<$list filter="[[$:/plugins/tgrosinger/tw5-checklist/Configuration]field:strike-checked[true]]">
```
This looks really cool, can you put a comment on the line above explaining what is happening?

tgrosinger reviewed (commented):
I haven't quite finished the review yet, but here is what I have so far. Looks really good @telmiger, thank you for putting so much work into this!

tgrosinger commented on js/checkbox-wikirule.js:
```diff
@@ -79,50 +113,80 @@ exports.parse = function() {
+        };
+
+        // Make a button to delete the item
+        // TODO: use svg bin icon, pattern as follows?
```
Can you please pull this TODO out of the code and open a new issue with this content?

telmiger replied:
If we can proceed with my newest version, then this is already obsolete/done.

tgrosinger commented on js/checkbox-wikirule.js:
```diff
@@ -79,50 +113,80 @@ exports.parse = function() {
+            tag: "span",
+            attributes: {
+                class: {type: "string", value: "checklist-vh"},
+// the next line does nothing, it works here but nowhere else :–(
```
What do you mean by this? Why leave the line if it doesn't work?

telmiger replied:
I thought this could be a clue for a more elegant solution using (translated) standard texts … but if you don’t know how to make this work either, then I will remove it.

tgrosinger replied:
Ah, no I am not familiar with textReference

tgrosinger commented on js/checklist.js:
```diff
@@ -56,6 +56,7 @@ CheckListWidget.prototype.execute = function() {
     this.makeChildWidgets();
     this.renderChildren(domNode);
 
+    //  If this is the clear all button … 
```
Move this down one line, into the each function

telmiger replied:
Got it. I have replaced it with /* add event listeners */ in the upcoming version – the following if-statement speaks for itself.

tgrosinger commented on js/checklist.js:
```diff
@@ -111,13 +113,15 @@ CheckListWidget.prototype.handleClearChecksEvent = function(event) {
 
 // When the user starts typing, change the pencil icon into a checkbox
 CheckListWidget.prototype.handleNewItemTypingEvent = function(event) {
```
Why don't we just remove this whole block then instead of leaving in commented-out code

telmiger replied:
Two reasons: 1) I wanted to let you know. 2) This function is called from somewhere and I did not have the time to clean up properly … will do so.

tgrosinger commented on tiddlers/stylesheet.tid:
```diff
@@ -1,64 +1,135 @@
+    width: 18px;
+    height: 18px;
+    border-radius: 5px;
+background-image: url('data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2211%22%20height%3D%2211%22%3E%0A%3Crect%20width%3D%221%22%20height%3D%227%22%20x%3D%225%22%20y%3D%222%22%20fill%3D%22%239B9B9B%22%2F%3E%0A%3Crect%20width%3D%227%22%20height%3D%221%22%20x%3D%222%22%20y%3D%225%22%20fill%3D%22%239B9B9B%22%2F%3E%0A%3C%2Fsvg%3E');
```
Is this just a small "+" icon? Can we just use some sort of unicode character instead?

telmiger replied:
We can, if you know a unicode character that looks like a + but is not announced by screenreaders. The recommended way is to use a background image and a button-label. With TW 5.1.14 we have a [more elegant way to create datauri](http://tiddlywiki.com/#datauri%20Macro) – would it be o.k. if the new version is TW 5.1.14+ only?

tgrosinger replied:
Let's wait to use that until 5.1.14 has been out a bit longer. Want to keep compatibility as flexible as possible

tgrosinger commented on tiddlers/stylesheet.tid:
```diff
@@ -1,64 +1,135 @@
-div.checklist-removeitem-icon {
-    bottom: 0;
-    color: #777;
+<$list filter="[[$:/plugins/tgrosinger/tw5-checklist/Configuration]field:strike-checked[true]]">
```
Does this just only include the surrounded body if the field is true?

telmiger replied:
Exactly, you can check the preview/rendered tiddler and you should see the result.

tgrosinger commented:
@telmiger Can you put up a an updated version of your demo when you get a chance? Thank you again!

telmiger commented:
The new version, in which I have tried to consider all your inputs, is online: https://tid.li/tw5/test/checklist.html – I hope you like it. 
If I find out how to update a PR, I will be happy to do so.

tgrosinger commented:
@telmiger the new demo looks awesome! Once you push the changes to the pull request I will take a final pass over it and then we can get it merged.

Please notice that I made a small change in the Usage.tid which has caused a merge conflict in your review.

tgrosinger commented:
@telmiger  I just checked out your branch to build the state that has been uploaded, but it is definitely missing changes that are reflected in the demo you uploaded. You should be able to just use a `git push` to upload a new version of your changes.

telmiger committed:
Reflect changes by tgrosinger

Make my PR conflict free

telmiger committed:
Optimize buttons, highlighting and colors

telmiger committed:
Better labeling and cleanup

Comments were transfered to new issues

telmiger committed:
Remove typing effect

The checkbox that replaced the pencil before the input field did not work in the sense of marking the newly entered item, so it was removed.

telmiger commented:
My master branch is updated. Should I send a new PR or can you proceed with this one here?

tgrosinger commented:
No need for a new PR, I can see the commits and will take a look at them as soon as I get a chance. Thank you!
